core/src/main/java/com/novoda/merlin/service/MerlinServiceBinder.java

```diff
@@ -70,16 +69,31 @@ private boolean connectionIsAvailable() {
         @Override
         public void onServiceConnected(ComponentName name, IBinder binder) {
             Logger.d("onServiceConnected");
-            merlinService = ((MerlinService.LocalBinder) binder).getService();
-            merlinService.setConnectListener(listenerHolder.connectListener);
-            merlinService.setDisconnectListener(listenerHolder.disconnectListener);
-            merlinService.setBindStatusListener(listenerHolder.bindListener);
-            merlinService.setHostname(endpoint, validator);
+            MerlinService.LocalBinder merlinServiceBinder = ((MerlinService.LocalBinder) binder);
+            ConnectivityManager connectivityManager = (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
+            ConnectivityChangesRegister connectivityChangesRegister = new ConnectivityChangesRegister(
+                    context,
+                    connectivityManager,
+                    new AndroidVersion()
+            );
+            NetworkStatusRetriever networkStatusRetriever = new NetworkStatusRetriever(MerlinsBeard.from(context));
+            HostPinger hostPinger = HostPinger.withCustomEndpointAndValidation(endpoint, validator);
+            ConnectivityChangesForwarder connectivityChangesForwarder = new ConnectivityChangesForwarder(
+                    networkStatusRetriever,
+                    listenerHolder.disconnectListener,
+                    listenerHolder.connectListener,
+                    listenerHolder.bindListener,
+                    hostPinger
+            );
+
+            merlinServiceBinder.setConnectivityChangesRegister(connectivityChangesRegister);
+            merlinServiceBinder.setConnectivityChangesForwarder(connectivityChangesForwarder);
+            merlinServiceBinder.onBindComplete();
         }
 
         @Override
         public void onServiceDisconnected(ComponentName name) {
-            merlinService = null;
+            // do nothing.
```

Comment:
Bind and forget, the objects are destroyed when the service is unbound. No need to reference the `Service` directly.